c/misra/src/rules/RULE-5-4/MacroIdentifiersNotDistinct.ql

```diff
@@ -0,0 +1,33 @@
+/**
+ * @id c/misra/macro-identifiers-not-distinct
+ * @name RULE-5-4: Macro identifiers shall be distinct
+ * @description Declaring multiple macros with the same name leads to undefined behaviour.
+ * @kind problem
+ * @precision very-high
+ * @problem.severity warning
+ * @tags external/misra/id/rule-5-4
+ *       correctness
+ *       maintainability
+ *       readability
+ *       external/misra/obligation/required
+ */
+
+import cpp
+import codingstandards.c.misra
+
+from Macro m, Macro m2
+where
+  not isExcluded(m, Declarations1Package::macroIdentifiersNotDistinctQuery()) and
+  not m = m2 and
+  (
+    //C99 states the first 63 characters of macro identifiers are significant
+    //C90 states the first 31 characters of macro identifiers are significant and is not currently considered by this rule
+    //ie an identifier differing on the 32nd character would be indistinct for C90 but distinct for C99
+    //and is currently not reported by this rule
+    if m.getName().length() >= 64
+    then m.getName().substring(0, 62) = m2.getName().substring(0, 62)
+    else m.getName() = m2.getName()
+  ) and
+  //reduce double report since both macros are in alert, arbitrary ordering
+  m.getLocation().getStartLine() >= m2.getLocation().getStartLine()
+select m, "Nondistinct macro identifer used " + m.getName() + " compared to $@.", m2, m2.getName()
```

Comment:
We shouldn't report macros with ifdef(n) guards around them, e.g.
```c
  #ifndef MY_MACRO
    #define MY_MACRO 1
  #endif
```

Reply:
can you suggest a testcase that makes a false positive from this case? I think since the DB only contains Macros that actually make it into source, the immediate case I came up with does not FP:

testcase:
```
#ifndef CONDITION_MACRO
#define MY_MACRO 1 // COMPLIANT
#else
#define MY_MACRO 2 // COMPLIANT
#endif

#ifndef MY_MACRO
#define MY_MACRO 1 // COMPLIANT
#endif
```
query:
```
from Macro m
select m , m.getName()
```
result:
```
+| test.c:12:1:12:18 | #define MY_MACRO 1 | MY_MACRO |
```
if MY_MACRO only ends up being defined once due to conditional compilation, there is no FP. if its a redefinition, well that is reported but I would say we arent really looking to guess about purposeful redefinition... like:
```
#define MY_MACRO 2 // COMPLIANT

#ifdef MY_MACRO
#define MY_MACRO 1 // NON_COMPLIANT
#endif
```
 unless that is the case you are suggesting to omit?

Reply:
Yes, I think you're right - it doesn't seem possible to get a false positive with an if(n)def. Thanks for checking!